docs: running more than one service locally (#1648)
lightbend/kalix-jvm-sdk: docs/src/modules/java-protobuf/pages/running-locally.adoc

= Run a service locally
:page-supergroup-java-scala: Language
include::ROOT:partial$include.adoc[]

Running a service locally is helpful to test and debug. The following sections provide commands for starting and stopping a single service locally.

== Prerequisites

In order to run your service locally, you'll need to have the following prerequisites:

* Docker {minimum_docker_version} or higher

> * Access to the `gcr.io/kalix-public` container registry

Reviewer: How do users know what this means, do they have access automatically, is it public, do they need to login with Kalix CLI? Clarifying that would be good.

Author: Good point. I don't know why we have it. This part is a copy of what we had on the main docs. 

So it's public and accessible to everyone. Users don't need to login to Kalix or whatsoever. 

Maybe the intention was to say that `gcr.io/kalix-public` should not be blocked by a firewall. 🤷

[TIP]
====

> The samples provided with the Kalix SDKs all have `docker-compose` files preconfigured to run your service locally. They also contain the configuration to start a local Google Pub/Sub emulator that the Kalix proxy will connect to when you make use of topic-based Eventing in your service.

Reviewer: Didn't we comment out the pub/sub from samples not using topics, or was that only something discussed?

Author: yes, I will remove this.

====

== Starting your service

To start your service locally, run the following command from the root of your project:

[.tabset]
Java::
+
--
[source, command line]
----
mvn kalix:runAll
----
--

Scala::
+
--
[source, command line]
----
sbt runAll
----
--

This command will start your Kalix service and a Kalix Proxy using the included `docker-compose.yml` file.

> If you prefer, you can also start `docker-compose` directly by running `docker-compose up` in one terminal and in another terminal start your Kalix service with:

Reviewer: ```suggestion
If you prefer, you can instead start `docker-compose` manually by running `docker-compose up` in one terminal and in another terminal start your Kalix service with:
```

[.tabset]
Java::
+
--
[source, command line]
----
mvn kalix:run
----
--

Scala::
+
--
[source, command line]
----
sbt run
----
--

== Invoking your service

> After you start the service it will accept invocations on `localhost:9000`. You can use gRPCurl to invoke your service.

Reviewer: ```suggestion
After you start the service it will accept invocations on `localhost:9000`. You can use [gRPCurl](https://github.com/fullstorydev/grpcurl) to invoke your service.
```

Author: Agree, but need asciidoc syntax.

```suggestion
After you start the service it will accept invocations on `localhost:9000`. You can use https://github.com/fullstorydev/grpcurl[gRPCurl] to invoke your service.
```

 --plaintext -d localhost:9000 \ 
  customer.api.CustomerService/Create
----
--

- The message payload in JSON format with `-d`
- The address of the proxy (using `plaintext` instead of TLS)
- The RPC procedure name which is deduced from the protobuf definition of the component you're calling.

Windows 10+::
+
--
[source,command line]
----
grpcurl ^ 
 -d '{"customer_id": "vip", "email": "vip@example.com", "name": "Very Important", "address": {"street": "Road 1", "city": "The Capital"}}' ^ 
 --plaintext -d localhost:9000 ^  
  customer.api.CustomerService/Create 
----
--

- The message payload in JSON format with `-d`
- The address of the proxy (using `plaintext` instead of TLS)
- The RPC procedure name which is deduced from the protobuf definition of the component you're calling.

== Shut down the service

Use `Ctrl+C` to shutdown the service. When stopping your service, it will also shutdown the Kalix Proxy container and any other container you have defined in `docker-compose.yml`. Unless you have chosen to start `docker-compose` manually in a separate terminal, in which case you will need to stop it manually as well. 

== Running multiple services locally

> A typical Kalix application is composed of one or more services deployed to the same Kalix project. When deployed under the same Kalix project, two different services can make xref:call-another-service.adoc[calls to each other] or xref:service-to-service.adoc[subscribe to each other event streams] by simply using their logical names.

Author: ```suggestion
A typical Kalix application is composed of one or more services deployed to the same Kalix project. When deployed under the same Kalix project, two different services can make xref:call-another-service.adoc[calls to each other] or xref:service-to-service.adoc[subscribe to each other's event streams] by simply using their logical names.
```

----

=== Customer Registry Subscriber Sample

On the other hand, in the *customer-registry-subscriber* we will use port 8081 and 9001 respectively to avoid port conflicts with the *customer-registry* service.

Moreover, since *customer-registry-subscriber* needs to subscribe to *customer-registry* and since we will be running it on our local machine, we need to _teach_ it where to find the *customer-registry* service.

This is done by passing an extra property (`kalix.dev-mode.service-port-mappings.customer-registry`) to its Kalix Proxy to let it create a mapping between the logical name *customer-registry* and the host and port where the *customer-registry* service is running.

Note that you need to add service port mappings to the Kalix Proxy configuration for the service that depends on the other service. Here, *customer-registry-subscriber* depends on *customer-registry*, therefore we add the service port mapping to the Kalix Proxy associated to *customer-registry-subscriber*.

[source,yml,indent=0]
----
include::java-protobuf:example$java-protobuf-eventsourced-customer-registry-subscriber/docker-compose.yml[tag=customer-registry-subscriber]
----

> With both services configured, we can start them independently by running  [.group-java]#`mvn kalix:runAll`# [.group-scala]#`sbt runAll`# in two disinct terminals. 

Reviewer: ```suggestion
With both services configured, we can start them independently by running  [.group-java]#`mvn kalix:runAll`# [.group-scala]#`sbt runAll`# in two separate terminals. 
```

At the root of the *customer-registry-subscriber* service, we can find an alternative `docker-compose` file called `docker-compose-integration.yml`. 

[source,yml,indent=0]
----
include::java:example$java-spring-eventsourced-customer-registry-subscriber/docker-compose-integration.yml[tag=customer-registry-docker]
----

But first we need to build a image for *customer-registry*. For local development, we don't need to publish it to a remote container registry. It suffice to build it locally. 

We can build the image by calling the following command at the root of  *customer-registry*:

[.tabset]
Java::
+
--
[source, command line]
----
mvn package docker:build
----
--

Scala::
+
--
[source, command line]
----
sbt docker:publishLocal
----
--

Next we can run *customer-registry-subscriber* and instead pass `docker-compose-integration.yml` to it. 

Run the following command at the root of *customer-registry-subscriber*:

[.tabset]
Java::
+
--
[source, command line]
----
mvn -Dkalix.dev-mode.docker-compose-file=docker-compose-integration.yml kalix:runAll
----
--

Scala::
+
--
[source, command line]
----

> sbt -Dkalix.dev-mode.docker-compose-file=docker-compose-integration.yml runAll

Author: This is not yet working. The mvn plugin passed further any `-D` properties, but the sbt plugin is not yet doing it.

----
--

This time,  [.group-java]#`mvn kalix:runAll`# [.group-scala]#`sbt runAll`# will start `docker-compose-integration.yml` instead. The *customer-registry* service and its companion Kalix Proxy will run alongside *customer-registry-subscriber* and its own Kalix Proxy. 

> Note that the configuraiton is exactly the same except that now we are running one single [.group-java]#`mvn kalix:runAll`# [.group-scala]#`sbt runAll`# command and the `docker-compose` file we are using contains all the dependencies required by *customer-registry-subscriber*.

Reviewer: ```suggestion
Note that the configuration is exactly the same except that now we are running one single [.group-java]#`mvn kalix:runAll`# [.group-scala]#`sbt runAll`# command and the `docker-compose` file we are using contains all the dependencies required by *customer-registry-subscriber*.
```

Author: `Vale` didn't pick that typo.
